Project/Source/Components/UI/ComponentVideo.cpp

```diff
@@ -0,0 +1,380 @@
+		[this]() { OpenVideoReader(App->resources->GetResource<ResourceVideo>(videoID)->GetAssetFilePath().c_str()); });
+
+	std::string removeButton = std::string(ICON_FA_TIMES "##") + "video";
+	if (ImGui::Button(removeButton.c_str())) {
+		if (videoID != 0) {
+			RemoveVideoResource();
+			App->resources->DecreaseReferenceCount(videoID);
+			videoID = 0;
+		}
+	}
+	ImGui::SameLine();
+	ImGui::TextUnformatted("Remove Video");
+
+	ResourceVideo* videoResource = App->resources->GetResource<ResourceVideo>(videoID);
+	if (videoResource != nullptr) {
+		ImGui::Separator();
+		ImGui::Checkbox("Flip Vertically", &verticalFlip);
+	}
+
+	//audioPlayer->OnEditorUpdate();
+}
+
+void ComponentVideo::Save(JsonValue jComponent) const {
+	jComponent[JSON_TAG_VIDEOID] = videoID;
+	jComponent[JSON_TAG_IS_FLIPPED] = verticalFlip;
+}
+
+void ComponentVideo::Load(JsonValue jComponent) {
+	videoID = jComponent[JSON_TAG_VIDEOID];
+	verticalFlip = jComponent[JSON_TAG_IS_FLIPPED];
+}
+
+void ComponentVideo::Draw(ComponentTransform2D* transform) {
+	if (imageUIProgram == nullptr) return;
+
+	glBindBuffer(GL_ARRAY_BUFFER, App->userInterface->GetQuadVBO());
+	glEnableVertexAttribArray(0);
+	glEnableVertexAttribArray(1);
+	glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 0, (void*) 0);
+	glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 0, (void*) ((sizeof(float) * 6 * 3)));
+	glUseProgram(imageUIProgram->program);
+
+	float4x4 modelMatrix = transform->GetGlobalScaledMatrix();
+	float4x4& proj = App->camera->GetProjectionMatrix();
+	float4x4& view = App->camera->GetViewMatrix();
+
+	if (App->userInterface->IsUsing2D()) {
+		proj = float4x4::D3DOrthoProjLH(-1, 1, App->renderer->GetViewportSize().x, App->renderer->GetViewportSize().y); //near plane. far plane, screen width, screen height
+		view = float4x4::identity;
+	}
+
+	ComponentCanvasRenderer* canvasRenderer = GetOwner().GetComponent<ComponentCanvasRenderer>();
+	if (canvasRenderer != nullptr) {
+		float factor = canvasRenderer->GetCanvasScreenFactor();
+		view = view * float4x4::Scale(factor, factor, factor);
+	}
+
+	glUniformMatrix4fv(imageUIProgram->viewLocation, 1, GL_TRUE, view.ptr());
+	glUniformMatrix4fv(imageUIProgram->projLocation, 1, GL_TRUE, proj.ptr());
+	glUniformMatrix4fv(imageUIProgram->modelLocation, 1, GL_TRUE, modelMatrix.ptr());
+
+	glActiveTexture(GL_TEXTURE0);
+	glUniform1i(imageUIProgram->diffuseLocation, 0);
+	glUniform4fv(imageUIProgram->inputColorLocation, 1, float4(1.f, 1.f, 1.f, 1.f).ptr());
+
+	// allocate memory and set texture data
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, frameWidth, frameHeight, 0, GL_RGBA, GL_UNSIGNED_BYTE, frameData);
+
+	glUniform1i(imageUIProgram->hasDiffuseLocation, 1);
+
+	glDrawArrays(GL_TRIANGLES, 0, 6);
+	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+	glDisable(GL_BLEND);
+}
+
+void ComponentVideo::OpenVideoReader(const char* filename) {
+	MSTimer timer;
+	timer.Start();
+
+	// Open video file
+	formatCtx = avformat_alloc_context();
```

Comment:
This might be the reason why you can't have multiple videos maybe? Idk what that function does but you are allocating a new context everytime

Reply:
AFAIK, the context is storing the general information about the video file itself, since it is initialized with a filename. Thus should be a problem to initialise more than one context for different video files?
Anyway I'm not 100% how everything is structured, so it could be possible. Thx for the suggestion! I'll take a look at it.

Reply:
OK, sounds weird to me that a context is created for each new video since the Contexts we've been using are created once and are shared amongst others. I don't know how the library uses the context, but it's usually problematic to have multiple contexts of one library structure (p.eg, Windows, imgui, opengl, etc.). Anyway, if it's not mentionted in the library, it probably doesn't matter much!

Reply:
Because it is not the context of the library, its the context of the video format.
Each format (.mp4, .flv, .mov...) has its own inner data structure. Libav reads it and stores this "format context"

```
formatCtx = avformat_alloc_context();                    // This is memory alloc referenced in formatCtx
if (avformat_open_input(&formatCtx, filename, nullptr, nullptr) != 0)  // This reads filename and stores the video data in formatCtx
```

Reply:
I'm with Miguel, it's weird to open a Context for each component. We can take a look at it with the audio iteration. For now,  keep this way

Reply:
Yes, this can be checked on a later PR